pyteal/ast/app.py

```diff
@@ -0,0 +1,185 @@
+from enum import Enum
+
+from ..types import TealType, require_type
+from ..ir import TealOp, Op
+from .leafexpr import LeafExpr
+from .expr import Expr
+from .maybe import MaybeValue
+from .int import EnumInt
+from .global_ import Global
+
+class OnComplete:
+    """An enum of values that Txn.on_completion() may return."""
+    NoOp = EnumInt("NoOp")
+    OptIn = EnumInt("OptIn")
+    CloseOut = EnumInt("CloseOut")
+    ClearState = EnumInt("ClearState")
+    UpdateApplication = EnumInt("UpdateApplication")
+    DeleteApplication = EnumInt("DeleteApplication")
+
+class AppField(Enum):
+    optedIn = (Op.app_opted_in, TealType.uint64)
+    localGet = (Op.app_local_get, TealType.anytype)
+    localGetEx = (Op.app_local_get_ex, TealType.none)
+    globalGet = (Op.app_global_get, TealType.anytype)
+    globalGetEx = (Op.app_global_get_ex, TealType.none)
+    localPut = (Op.app_local_put, TealType.none)
+    globalPut = (Op.app_global_put, TealType.none)
+    localDel = (Op.app_local_del, TealType.none)
+    globalDel = (Op.app_global_del, TealType.none)
+
+    def __init__(self, op: Op, type: TealType) -> None:
+        self.op = op
+        self.ret_type = type
+    
+    def get_op(self) -> Op:
+        return self.op
+    
+    def type_of(self) -> TealType:
+        return self.ret_type
+
+class App(LeafExpr):
+    """An expression related to applications."""
+
+    def __init__(self, field:AppField, args) -> None:
+        self.field = field
+        self.args = args
+
+    def __str__(self):
+        ret_str = "({}".format(self.field.get_op())
+        for a in self.args:
+            ret_str += " " + a.__str__()
+        ret_str += ")"
+        return ret_str
+
+    def __teal__(self):
+        teal = []
+        for arg in self.args:
+            teal += arg.__teal__()
+        teal += [TealOp(self.field.get_op())]
+        return teal
+
+    def type_of(self):
+        return self.field.type_of()
+
+    @classmethod
+    def id(cls):
+        """Get the ID of the current running application.
+        
+        This is the same as Global.current_application_id().
+        """
+        return Global.current_application_id()
+
+    @classmethod
+    def optedIn(cls, account: Expr, app: Expr):
+        """Check if an account has opted in for an application.
+
+        Args:
+            account: An index into Txn.Accounts that corresponds to the account to check. Must
+            evaluate to uint64.
+            app: The ID of the application being checked. Must evaluate to uint64.
+        """
+        require_type(account.type_of(), TealType.uint64)
+        require_type(app.type_of(), TealType.uint64)
+        return cls(AppField.optedIn, [account, app])
+    
+    @classmethod
+    def localGet(cls, account: Expr, key: Expr):
```

Comment:
I don't know if this is a good idea, but I wonder if we can take advantage of `__getitem__`, `__setitem__`, and `__delitem__` overloading here.

Reply:
I have considered using `__getitem__`, but I believe `__setitem__` and `__delitem__` are not able to return anything so we wouldn't be able to use them as a drop in replacement for these AST functions. They would be great to have in a higher level approach that generates an AST though.

One alternative interface I have thought about is `Global.store(k, v)`, `Global.get(k)`, `Txn.accounts[0].store(k, v)`, `Txn.accounts[0].get(k)`. Some open questions with this are:
* Should `Txn.accounts[0]` still represent the address when used in an AST? Or should we have `Txn.accounts[0].address()`?
* How can I manipulate the sender's local state?
* The indexes in this array approach are one less than the actual runtime indexes would need to be. So should we only allow compile-time ints to be used? Or should `Txn.accounts[Int(1)]` be allowed, and what would it resolve to?

Reply:
> __setitem__ and __delitem__ are not able to return anything so we wouldn't be able to use them as a drop in replacement for these AST functions

I don't think `app_*_put` and `app_*_del` place anything on the stack; were you thinking that there might be a value returned by those that is of interested to PyTEAL?

Reply:
> Should `Txn.accounts[0]` still represent the address when used in an AST? Or should we have `Txn.accounts[0].address()`?

> How can I manipulate the sender's local state?

I didn't think this idea through completely, but could you alternatively have `Txn.accountStorage[0]`?

Reply:
I do like the idea of "calling" something to get the local storage corresponding to an account--as you point out, we do a bit of translation (subtract 1) under the covers

Reply:
> I don't think `app_*_put` and `app_*_del` place anything on the stack; were you thinking that there might be a value returned by those that is of interest to PyTEAL?

With the current way we build the AST, constructing an expression must return an AST for that expression, regardless of whether it actually places something on the stack. For example,
```python
program = Seq([
    Assert(App.global["status"] == Int(0)), # this is ok
    App.global["status"] = Int(1), # doesn't work because `__setitem__` can't return a value
    del App.global["status"] # doesn't work becasue `__delitem__` can't return a value
])
```
We would have to use contexts or another front end in order to successfully use `__setitem__` and friends. I am definitely open to that, just not in this release 😅

Reply:
> I didn't think this idea through completely, but could you alternatively have `Txn.accountStorage[0]`?

I will look into this today.

Reply:
I've combined the `Txn.accountStorage` and `App.global[key].set/delete` proposals into this: https://gist.github.com/jasonpaulos/28844a66ec0aae10c3629a0fdc2ae512. I've implemented part of the backend for it and I think it works pretty well.

Reply:
Oh, interesting. I do think I like the current approach in this PR, but this seems like it's getting pretty subjective. @algorandskiy any thoughts here?

Reply:
We've decided to wait on any change here for this release.